### DISCUSSION TITLE
Which Ammonite version, and combining with other com.lihaoyi dependencies

### DISCUSSION THREAD

**Discussion by bobismijnnaam** [lihaoyi/Ammonite#1382] [Q&A]
I'm a bit lost. I want to use the ammonite library in my project, as well as some other com.lihaoyi libraries (most prominently: cask, mainargs, os-lib). But I'm running into some strange problems whenever I include ammonite as a dependency. One thing that goes wrong sometimes is that it doesn't seem jive with my scala version, and I get very arcane type errors (this happens when I use scala 3.0.0, with a clean build):

```
scala: error while loading Typeable$package$,
class file scala/reflect/Typeable$package.class is broken, reading aborted with class dotty.tools.tasty.UnpickleException
TASTy signature has wrong version.
 expected: {majorVersion: 28, minorVersion: 0}
 found   : {majorVersion: 28, minorVersion: 1}

This TASTy file was produced by a more recent, forwards incompatible release.
To read this TASTy file, please upgrade your tooling.
The TASTy file was produced by Scala 3.1.3-bin-nonbootstrapped.
```

Then I increase the scala version to say 3.3.0, increasing the ammonite version in the process, and then a dependency conflict occurs between e.g. cask and ammonite, about a transitive dependency (usually geny, given the various times I've run into this now, with various permutations of libraries) being included with different version. E.g., using this sbt code:

```
name := "xyz"
version := "0.1"
scalaVersion := "3.3.0"

libraryDependencies += "com.lihaoyi" %% "ammonite" % "3.0.0-M0-56-1bcbe7f6" cross CrossVersion.full
libraryDependencies += "com.lihaoyi" %% "cask" % "0.9.1"
```

I get the following sbt when running `sbt compile`:

```
[info] welcome to sbt 1.9.7 (Private Build Java 17.0.8.1)
[info] loading project definition from /home/bobe/Downloads/xyz/project
[info] loading settings for project xyz from build.sbt ...
[info] set current project to xyz (in build file:/home/bobe/Downloads/xyz/)
[error] stack trace is suppressed; run last update for the full output
[error] (update) found version conflict(s) in library dependencies; some are suspected to be binary incompatible:
[error]
[error]         * com.lihaoyi:geny_2.13:1.0.0 (early-semver) is selected over 0.6.10
[error]             +- com.lihaoyi:fastparse_2.13:3.0.0                   (depends on 1.0.0)
[error]             +- com.lihaoyi:ammonite-cross-3.3.0-repl-api_2.13.11:3.0.0-M0-56-1bcbe7f6 (depends on 1.0.0)
[error]             +- com.lihaoyi:upickle-core_2.13:3.0.0                (depends on 1.0.0)
[error]             +- com.lihaoyi:os-lib_2.13:0.9.0                      (depends on 1.0.0)
[error]             +- com.lihaoyi:requests_2.13:0.7.0                    (depends on 0.6.10)
[error]
[error]
[error] this can be overridden using libraryDependencySchemes or evictionErrorLevel
[error] Total time: 0 s, completed Nov 6, 2023, 5:12:04 PM
```

What is the general way to have a project with multiple com.lihaoyi dependencies, given that I want to be on a recent scala version, preferably with recent/stable versions of those libraries? Would this problem be resolved if I used mill instead? Or can I maybe tell sbt to use this or that version, hopefully without breaking the libraries in turn?

**Comment by lefou** [comment]
Looks like your version of `requests-scala` is too old. 

> ```
> [error]             +- com.lihaoyi:requests_2.13:0.7.0                    (depends on 0.6.10)
> ```

You should update to `0.8.0` which seems to be compatible with geny `1.0.0`.

**Reply by bobismijnnaam** [reply]
Thanks for the quick response! This error occurs even with the minimal sbt example I posted, which has only cask and ammonite as dependencies. Does you mean my environment somehow prefers an old version of requests, or do you mean I should force either ammonite or cask to use a newer version of requests?

**Reply by lefou** [reply]
I can't tell you why you end up with requests `0.7.0` on your classpath, but sbt is telling you, it's transitive dependency on geny `0.6.10` isn't compatible to `1.0.0`. Either you exclude it or explicitly force the newer version. You could also try to inspect the whole transitive dependency tree (or post it here). I'm not an sbt user, so I can't tell you *how* to do it.

**Reply by bobismijnnaam** [reply]
Okay, thanks! I will investigate.
